### DISCUSSION TITLE
랭킹 캐싱 처리

### DISCUSSION THREAD

**Discussion by kimminkyeu** [Kernel360/F2-BAGUNI#746] [이슈 / 버그]
2. 랭킹 정보 자체는 집계 데이터가 많아지면 `랭킹 모듈 내부`에 캐싱이 필요 하다.
    1. 캐싱 필요
        
        랭킹을 페이지 조회 할 때마다 획득해 올 텐데. 
        
        그럼 그 획득은 같은 데이터 이거나 큰 변화가 없는 데이터이거나 사용자에게 반드시 최신을 보여줄 필요가 없는 데이터니까.
        
        이때 캐싱이 필요. 프론트 입장에서도. 리프레시 할때마다 랭킹 연산이 들어갈 필요가 전혀 없다.
        
    2. 구현 방식
        
        상원님이 로컬 캐시 라이브러리 (`카페인`) 써서 진행 중
        
        - 레디스르 쓰지 않는 이유는 레디스까지 쓸 만큼 분산환경도 아니고, 간단히 처리할 수 있으니까.
            
            또 CPU 점유율이 높아져 EC2에 올릴 경우 비용이 증가할 것이라 판단
            
        - 직접 로컬 캐시를 구현하지 않는 이유는, 일단 빠르게 적용할 수 있어서.

**Comment by sangwonsheep** [comment]
https://techblog.uplus.co.kr/%EB%A1%9C%EC%BB%AC-%EC%BA%90%EC%8B%9C-%EC%84%A0%ED%83%9D%ED%95%98%EA%B8%B0-e394202d5c87

![스크린샷 2024-12-11 오후 7 11 54](https://github.com/user-attachments/assets/51f399d8-cfad-4638-b91a-621ff5a32d15)

## 사용 이유
1. 분산 환경이면 Redis를 두어서 여러 서버에서 하나의 캐시를 사용하면 되지만, 우리는 분산 환경이 아니므로 속도가 더 빠른 로컬 캐시 채택하였음.
2. 로컬 캐시 중에 가장 성능이 좋음. 캐싱의 가장 큰 목적은 Read이므로 Read 성능도 가장 뛰어난 Caffeine 사용

**Comment by kimminkyeu** [comment]
- 로컬 캐시를 쓴 이유?
- 다른 라이브러리들이 많은데 와이?
- concurrent hash map과 뭐가 다름?
    - **time based conviction**
- why not redis?
- https://www.baeldung.com/java-caching-caffeine
- https://medium.com/naverfinancial/%EB%8B%88%EB%93%A4%EC%9D%B4-caffeine-%EB%A7%9B%EC%9D%84-%EC%95%8C%EC%95%84-f02f868a6192
- https://techblog.uplus.co.kr/%EB%A1%9C%EC%BB%AC-%EC%BA%90%EC%8B%9C-%EC%84%A0%ED%83%9D%ED%95%98%EA%B8%B0-e394202d5c87
